akka-http-core/src/main/scala/akka/http/engine/server/HttpServerBluePrint.scala

```diff
@@ -240,4 +266,75 @@ private[http] object HttpServerBluePrint {
         case _ ⇒ ctx.fail(error)
       }
   }
+
+  case class WebsocketSetup(
+    flow: Flow[ByteString, ByteString, Any],
+    publisherKey: StreamUtils.ReadableCell[Publisher[FramePart]],
+    subscriberKey: StreamUtils.ReadableCell[Subscriber[FramePart]],
+    switchSource: Source[SwitchToWebsocketToken.type, Any]) extends WebsocketSwitch {
+    @volatile var switchToWebsocketRef: Option[ActorRef] = None
+
+    def switchToWebsocket(handlerFlow: Flow[FramePart, FramePart, Any])(implicit mat: FlowMaterializer): Unit = {
+      // 1. fill processing hole in the websocket pipeline with user-provided handler
+      Source(publisherKey.value)
+        .via(handlerFlow)
+        .to(Sink(subscriberKey.value))
+        .run()
+
+      // FIXME: is there a race condition between 1 and 2?
+
+      // 2. flip the switch
+      switchToWebsocketRef.get ! TokenSourceActor.Trigger
+    }
+  }
+  def websocketPipeline: WebsocketSetup = {
+    val sinkCell = new StreamUtils.OneTimeWriteCell[Publisher[FramePart]]
+    val sourceCell = new StreamUtils.OneTimeWriteCell[Subscriber[FramePart]]
+
+    val sink = StreamUtils.oneTimePublisherSink[FramePart](sinkCell, "frameHandler.in")
+    val source = StreamUtils.oneTimeSubscriberSource[FramePart](sourceCell, "frameHandler.out")
+
+    val flow =
+      Flow[ByteString]
+        .transform[FramePart](() ⇒ new FrameEventParser)
+        .via(StreamUtils.sinkSourceFlow(sink, source))
+        .transform(() ⇒ new FrameEventRenderer)
+
+    lazy val setup = WebsocketSetup(flow, sinkCell, sourceCell, switchToWebsocketSource)
+    lazy val switchToWebsocketSource: Source[SwitchToWebsocketToken.type, ActorRef] =
+      Source.actorPublisher[SwitchToWebsocketToken.type] {
+        Props {
+          val actor = new TokenSourceActor(SwitchToWebsocketToken)
+          setup.switchToWebsocketRef = Some(actor.context.self)
+          actor
+        }
```

Comment:
This should work, but why not use `actor.self`? OTOH this whole dance should be avoidable by using an AsyncStage instead of the sinkSourceFlow where the materialized value allows sending in the real Flow to be materialized when required. The problem right now is that the AsyncStage would have to manually implement the Publisher and Subscriber interfaces because we have not yet ported the InputBunch / OutputBunch from ActorProcessor to the new scheme—the plan was to do that after 1.0 because it is an internal improvement.